dvc/stage/__init__.py

```diff
@@ -314,7 +314,8 @@ def _changed_deps(
             if status:
                 if allow_missing and status[str(dep)] == "deleted":
                     if upstream and any(
-                        dep.fs_path == out.fs_path and dep.hash_info != out.hash_info
+                        dep.fs_path == out.fs_path
+                        and dep.hash_info.value != out.hash_info.value
```

Comment:
We should not look into  `hash_info.value`. 2.0 and 3.0 hashes are not compatible.

Reply:
Yes, I know the theory:

https://github.com/iterative/dvc/issues/9818#issuecomment-1674867181

Do we care in practice (today, not if we introduce different hashing algorithm)? 

I mean, isn't the same chance of collision as today with 2 different `3.0` hashes?

Reply:
Even without considering the collision, if I am understanding the branch of this code correctly, there is a chance of hashes being mismatched and returning `modified`, no?

Reply:
@daavoo hash_info should be compared as a whole, comparing only values without name looks like a hack. Could you elaborate on why you need to do it this way?

Reply:
I guess new output is getting hashed as `3.x` and `deps` is hashed as `2.x` format?

Reply:
Seems like it would be better to just ask to upgrade (e.g. with `dvc commit`) https://github.com/iterative/dvc/issues/9818#issuecomment-1674896843 Otherwise we have to compute both hashes (which is undesirable) or hack like it is done in this PR, which makes it straight up error-prone.

Reply:
> there is a chance of hashes being mismatched and returning modified, no?

Not sure I follow, sorry. 
This is what happens before the P.R: an equivalent folder is mismatched because of `hash_info.name` but only when the content is missing. If you pull the data, the same `.dvc` and `dvc.lock` report it as not modified